File: go-backend-interview-prep/05_errors_and_testing/02_wrap_errors_and_unwrap/main.py
"""Wrap errors and unwrap -- Python equivalent of the Go example."""


# --- Custom exception types ---
class NotFoundError(Exception):
    pass


class PermissionError_(Exception):
    """Underscore to avoid shadowing built-in PermissionError."""
    def __init__(self, user: str, action: str):
        self.user = user
        self.action = action
        super().__init__(f"user {user!r} cannot {action}")


# --- Layered functions that chain exceptions ---

def find_record(record_id: int) -> None:
    if record_id <= 0:
        raise NotFoundError("not found")
    if record_id == 99:
        raise PermissionError_(user="guest", action="read")


def get_profile(record_id: int) -> None:
    try:
        find_record(record_id)
    except Exception as e:
        raise RuntimeError(f"get_profile(id={record_id})") from e


def handle_request(record_id: int) -> None:
    try:
        get_profile(record_id)
    except Exception as e:
        raise RuntimeError("handle_request") from e


def main() -> None:
    # --- Example 1: isinstance walks the __cause__ chain ---
    try:
        handle_request(0)
    except Exception as e:
        print("Error:", e)
        # Walk the chain manually
        cause = e.__cause__
        while cause:
            if isinstance(cause, NotFoundError):
                print("Is NotFoundError? True")
                break
            cause = cause.__cause__
        else:
            print("Is NotFoundError? False")

    # --- Example 2: extract typed exception from chain ---
    try:
        handle_request(99)
    except Exception as e:
        print("\nError:", e)
        cause = e.__cause__
        while cause:
            if isinstance(cause, PermissionError_):
                print(f"Permission denied: user={cause.user} action={cause.action}")
                break
            cause = cause.__cause__

    # --- Example 3: raise...from preserves chain ---
    print()
    base = ValueError("disk full")
    try:
        raise RuntimeError("save file") from base
    except RuntimeError as e:
        print("with 'from' preserves cause:", e.__cause__ is base)

    # Without 'from' -- no __cause__
    try:
        raise RuntimeError(f"save file: {base}")
    except RuntimeError as e:
        print("without 'from' loses cause:", e.__cause__ is None)


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/05_errors_and_testing/02_wrap_errors_and_unwrap/main.go
package main

import (
	"errors"
	"fmt"
)

// --- Sentinel error ---
var ErrNotFound = errors.New("not found")

// --- Custom error type ---
type PermissionError struct {
	User   string
	Action string
}

func (e *PermissionError) Error() string {
	return fmt.Sprintf("user %q cannot %s", e.User, e.Action)
}

// --- Layered functions that wrap errors ---

func findRecord(id int) error {
	if id <= 0 {
		return ErrNotFound
	}
	if id == 99 {
		return &PermissionError{User: "guest", Action: "read"}
	}
	return nil
}

func getProfile(id int) error {
	err := findRecord(id)
	if err != nil {
		return fmt.Errorf("getProfile(id=%d): %w", id, err)
	}
	return nil
}

func handleRequest(id int) error {
	err := getProfile(id)
	if err != nil {
		return fmt.Errorf("handleRequest: %w", err)
	}
	return nil
}

func main() {
	// --- Example 1: errors.Is walks the chain ---
	err := handleRequest(0)
	fmt.Println("Error:", err)
	fmt.Println("Is ErrNotFound?", errors.Is(err, ErrNotFound))

	// --- Example 2: errors.As extracts a typed error ---
	err = handleRequest(99)
	fmt.Println("\nError:", err)

	var permErr *PermissionError
	if errors.As(err, &permErr) {
		fmt.Printf("Permission denied: user=%s action=%s\n", permErr.User, permErr.Action)
	}

	// --- Example 3: %v vs %w ---
	base := errors.New("disk full")
	withW := fmt.Errorf("save file: %w", base)
	withV := fmt.Errorf("save file: %v", base)

	fmt.Println("\n%w preserves chain:", errors.Is(withW, base))
	fmt.Println("%v breaks chain:   ", errors.Is(withV, base))
}
